nexus/deployment/src/blueprint_builder.rs

```diff
@@ -420,10 +357,109 @@ impl<'a> BlueprintBuilder<'a> {
     }
 }
 
+/// Helper for working with sets of zones on each sled
+///
+/// Tracking the set of zones is slightly non-trivial because we need to bump
+/// the per-sled generation number iff the zones are changed.  So we need to
+/// keep track of whether we've changed the zones relative to the parent
+/// blueprint.  We do this by keeping a copy of any `OmicronZonesConfig` that
+/// we've changed and a _reference_ to the parent blueprint's zones.  This
+/// struct makes it easy for callers iterate over the right set of zones.
+struct BlueprintZones<'a> {
+    changed_zones: BTreeMap<Uuid, OmicronZonesConfig>,
+    parent_zones: &'a BTreeMap<Uuid, OmicronZonesConfig>,
+}
+
+impl<'a> BlueprintZones<'a> {
+    pub fn new(parent_blueprint: &'a Blueprint) -> BlueprintZones {
+        BlueprintZones {
+            changed_zones: BTreeMap::new(),
+            parent_zones: &parent_blueprint.omicron_zones,
+        }
+    }
+
+    /// Returns a mutable reference to a sled's Omicron zones *because* we're
+    /// going to change them.  It's essential that the caller _does_ change them
+    /// because we will have bumped the generation number and we don't want to
+    /// do that if no changes are being made.
```

Comment:
I can think of some ways we could try to enforce this in the API, but they'd be worse than what you have, so I have a question instead. By "essential" here you mean "the control plane will end up doing unnecessary work if we bump the generation without making any changes", right? It's not essential from a correctness point of view?

Reply:
I think that's right.  And yeah, same: I thought about just tracking the zone list here, and figuring out whether it's changed at the end, and bumping the generation if so, but it didn't seem worth the complexity given the cost of getting this wrong.